### PR TITLE
Add new reduce options and autograd support for scatter_reduce

### PR DESCRIPTION
**Pull request by mikaylagawarecki**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #71256
* #70750

Differential Revision: [D33712802](https://our.internmc.facebook.com/intern/diff/D33712802)

### CODE REVIEW

File: aten/src/ATen/native/TensorAdvancedIndexing.cpp [hunk 1]
```diff
@@ -1314,21 +1315,47 @@ TORCH_IMPL_FUNC(scatter_reduce_structured_cpu)
...
+    // documentation specifies that the elements in out should be considered in the computation(?)
+    // fill with different things depending on mode
+    // but how to tell whether out Tensor was initialized in meta or is a genuine out tensor???
+    // doesn't seem to be possible to initialize values of out tensor in TORCH_META_FUNC
+    if (reduce == "sum" || reduce == "mean") {
+      out.fill_(0);
+    } else if (reduce == "prod") {
+      out.fill_(1);
+    } else if (reduce == "amax") {
+      out.fill_(INT_MIN);
```
Comment (cpuhrsch) [inline aten/src/ATen/native/TensorAdvancedIndexing.cpp:1339]: You should be able to use ```std::numeric_limits<scalar_t>::max()```

**Review by cpuhrsch — APPROVED** [review]
Looks great!

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712802).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712802).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712802).

**Comment by mikaylagawarecki** [comment]
@mikaylagawarecki has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33712802).

**Pull request closed without merging**